**Getting the dataset from the url and loading it**

In [1]:
import pandas as pd
url='https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csv'
df = pd.read_csv(url)

# Defining our input and output attributes as x and y respectively

In [2]:
X=df.drop(['species'],axis=1)
y=df['species']

# Splitting the data in to test and train

In [3]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.4, random_state=6)

# knn at k=3

In [4]:

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
KNN_model = KNeighborsClassifier(n_neighbors=3)
y_predi = KNN_model.fit(X_train, y_train)
y_predi=KNN_model.predict(X_test)
a=1-accuracy_score(y_test, y_predi)
# Check accuracy score 

print('ACCURACY  =  : {0:0.4f}'. format(accuracy_score(y_test, y_predi)))


ACCURACY  =  : 0.9667


**# observation:**

 At k=3 we got 96% testing accuracy 

**prediction at different values of k**

In [5]:
l=[1,5,7,11,15,27,59]
for i in l:
  KNN_model = KNeighborsClassifier(n_neighbors=i)
  y_predi = KNN_model.fit(X_train, y_train)
  y_predi=KNN_model.predict(X_test)
  a=1-accuracy_score(y_test, y_predi)
  # Check accuracy score 
  print('n_neighbors =',i)
  print('ACCURACY  =  : {0:0.4f}'. format(accuracy_score(y_test, y_predi)))
  print('----------------------------------------------------------------')


n_neighbors = 1
ACCURACY  =  : 0.9500
----------------------------------------------------------------
n_neighbors = 5
ACCURACY  =  : 0.9833
----------------------------------------------------------------
n_neighbors = 7
ACCURACY  =  : 0.9667
----------------------------------------------------------------
n_neighbors = 11
ACCURACY  =  : 0.9667
----------------------------------------------------------------
n_neighbors = 15
ACCURACY  =  : 0.9333
----------------------------------------------------------------
n_neighbors = 27
ACCURACY  =  : 0.9167
----------------------------------------------------------------
n_neighbors = 59
ACCURACY  =  : 0.8167
----------------------------------------------------------------


**# Observation:**

1:When k=5 we got the best accuracy i.e:98.33%

2:when k= 7 or k=11 we got the second best accuracy i.e 96.667%

3:When k= 1 we got the third best accuracy i.e 95.33%

4:When k= 15 we got the fourth best accuracy i.e 93.33%

5:When k=27 we got the fifth best accuracy i.e 91.67%

6:Worst accuracy we got is 81.67% when k=59

**Single feature prediction**

In [6]:
petal_len=X['petal_length']
sepal_len=X['sepal_length']
petal_wid=X['petal_width']
sepal_wid=X['sepal_width']

**test train split for petal length feature**

In [7]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(petal_len,y, test_size=0.4, random_state=6)

**KNN at k=3 for petal_length**

In [8]:

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
KNN_model = KNeighborsClassifier(n_neighbors=3)
y_predi = KNN_model.fit(X_train.values.reshape(-1,1), y_train)
y_predi=KNN_model.predict(X_test.values.reshape(-1,1))
a=1-accuracy_score(y_test, y_predi)
# Check accuracy score 

print('ACCURACY  =  : {0:0.4f}'. format(accuracy_score(y_test, y_predi)))


ACCURACY  =  : 0.9333


**test train split for sepal length feature**

In [9]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(sepal_len,y, test_size=0.4, random_state=6)

**KNN at k=3 for sepal_length**

In [10]:

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
KNN_model = KNeighborsClassifier(n_neighbors=3)
y_predi = KNN_model.fit(X_train.values.reshape(-1,1), y_train)
y_predi=KNN_model.predict(X_test.values.reshape(-1,1))
a=1-accuracy_score(y_test, y_predi)
# Check accuracy score 

print('ACCURACY  =  : {0:0.4f}'. format(accuracy_score(y_test, y_predi)))


ACCURACY  =  : 0.7167


**test train split for petal_width feature**

**KNN at k=3 for petal_width**

In [11]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(petal_wid,y, test_size=0.4, random_state=6)

In [12]:

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
KNN_model = KNeighborsClassifier(n_neighbors=3)
y_predi = KNN_model.fit(X_train.values.reshape(-1,1), y_train)
y_predi=KNN_model.predict(X_test.values.reshape(-1,1))
a=1-accuracy_score(y_test, y_predi)
# Check accuracy score 

print('ACCURACY  =  : {0:0.4f}'. format(accuracy_score(y_test, y_predi)))


ACCURACY  =  : 0.9500


**test train split for sepal_width feature**

In [13]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(sepal_wid,y, test_size=0.4, random_state=6)

**KNN at k=3 for sepal_width**

In [14]:

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
KNN_model = KNeighborsClassifier(n_neighbors=3)
y_predi = KNN_model.fit(X_train.values.reshape(-1,1), y_train)
y_predi=KNN_model.predict(X_test.values.reshape(-1,1))
a=1-accuracy_score(y_test, y_predi)
# Check accuracy score 

print('ACCURACY  =  : {0:0.4f}'. format(accuracy_score(y_test, y_predi)))


ACCURACY  =  : 0.5667


**Observation:**

1:The best accuracy we got for petal_width feature i.e 95%

2:Second best accuracy we got for petal_length feature i.e 93.33%

3:third best accuracy we got for sepal_length feature i.e 71.67%

4:Worst accuracy we got using sepal_width feature i.e 56.67%

**Two feature prediction**

In [15]:
one=X[['sepal_length','sepal_width']]
two=X[['sepal_length','petal_length']]
three=X[['sepal_length','petal_width']]
four=X[['sepal_width','petal_length']]
five=X[['sepal_width','petal_width']]
six=X[['petal_length','petal_width']]


**test train split for sepal length and sepal_width feature**

In [16]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(one,y, test_size=0.4, random_state=6)

# KNN at k=3 for sepal length and sepal_width

In [17]:

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
KNN_model = KNeighborsClassifier(n_neighbors=3)
y_predi = KNN_model.fit(X_train, y_train)
y_predi=KNN_model.predict(X_test)
a=1-accuracy_score(y_test, y_predi)
# Check accuracy score 

print('ACCURACY  =  : {0:0.4f}'. format(accuracy_score(y_test, y_predi)))


ACCURACY  =  : 0.8167


**test train split for sepal length and petal length feature**

# KNN at k=3 for  sepal length and petal length

In [18]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(two,y, test_size=0.4, random_state=6)

In [19]:

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
KNN_model = KNeighborsClassifier(n_neighbors=3)
y_predi = KNN_model.fit(X_train, y_train)
y_predi=KNN_model.predict(X_test)
a=1-accuracy_score(y_test, y_predi)
# Check accuracy score 

print('ACCURACY  =  : {0:0.4f}'. format(accuracy_score(y_test, y_predi)))


ACCURACY  =  : 0.9833


**test train split for sepal length and petal width feature**

# KNN at k=3 for sepal length and petal width 

In [20]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(three,y, test_size=0.4, random_state=6)

In [21]:

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
KNN_model = KNeighborsClassifier(n_neighbors=3)
y_predi = KNN_model.fit(X_train, y_train)
y_predi=KNN_model.predict(X_test)
a=1-accuracy_score(y_test, y_predi)
# Check accuracy score 

print('ACCURACY  =  : {0:0.4f}'. format(accuracy_score(y_test, y_predi)))


ACCURACY  =  : 0.9500


**test train split for sepal width and petal length feature**

In [22]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(four,y, test_size=0.4, random_state=6)

# KNN at k=3 for  sepal width and petal length 

In [23]:

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
KNN_model = KNeighborsClassifier(n_neighbors=3)
y_predi = KNN_model.fit(X_train, y_train)
y_predi=KNN_model.predict(X_test)
a=1-accuracy_score(y_test, y_predi)
# Check accuracy score 

print('ACCURACY  =  : {0:0.4f}'. format(accuracy_score(y_test, y_predi)))


ACCURACY  =  : 0.9500


**test train split for sepal width and petal width feature**

In [24]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(five,y, test_size=0.4, random_state=6)

# KNN at k=3 for sepal width and petal width

In [25]:

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
KNN_model = KNeighborsClassifier(n_neighbors=3)
y_predi = KNN_model.fit(X_train, y_train)
y_predi=KNN_model.predict(X_test)
a=1-accuracy_score(y_test, y_predi)
# Check accuracy score 

print('ACCURACY  =  : {0:0.4f}'. format(accuracy_score(y_test, y_predi)))


ACCURACY  =  : 0.9500


**test train split for petal length and petal_width feature**

In [26]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(six,y, test_size=0.4, random_state=6)

# KNN at k=3 for petal length and petal width

In [27]:

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
KNN_model = KNeighborsClassifier(n_neighbors=3)
y_predi = KNN_model.fit(X_train, y_train)
y_predi=KNN_model.predict(X_test)
a=1-accuracy_score(y_test, y_predi)
# Check accuracy score 

print('ACCURACY  =  : {0:0.4f}'. format(accuracy_score(y_test, y_predi)))


ACCURACY  =  : 0.9667


# Observation:

1:The first best accuracy we got with features sepal length and petal length i.e 98.33%

2:Second best accuracy we got with features petal length and petal width i.e 96.67%

3:third best accuracy we got with features sepal width and petal width i.e 95%

4:fourth best accuracy we got with features sepal width and petal length i.e 95%   

5:fifth best accuracy we got with features sepal length and petal width i.e 95%

6:Worst accuracy we got using features sepal length and sepal_width i.e 81.67%